File: windows/x64/cpp/test/test_omnihand_dex_umi.cc
// Copyright (c) 2025, Agibot Co., Ltd.
// OmniHand 2025 SDK is licensed under Mulan PSL v2.

#include <gtest/gtest.h>
#include "omnihand/omnihand_dex_umi.h"
#include <memory>
#include <vector>
#include <iostream>
#include <iomanip>
#include <thread>
#include <chrono>
#include <string>

// Global variable to store device type from command line argument
static std::string g_device_type = "zlgcan";  // Default: zlgcan

// Helper function to get device type
static std::string GetDeviceType() {
  return g_device_type;
}

class OmniHandDexUMITest : public ::testing::Test {
 protected:
  void SetUp() override {
    // Create hand instance for testing based on device type
    std::string device_type = GetDeviceType();
    if (device_type == "hcan") {
      hand_ = agilink::omnihand::OmniHandDexUMI::createHandByHcan(
          agilink::omnihand::HandType::LEFT,        // hand_type: left hand
          1,                       // hand_device_id: hand device ID
          0,                       // canfd_device_id: USB CANFD adapter device index
          0                        // canfd_channel_id: CAN channel index (0=can0, 1=can1)
      );
    } else {  // default: zlgcan
      hand_ = agilink::omnihand::OmniHandDexUMI::createHandByZlgcan(
          agilink::omnihand::HandType::LEFT,        // hand_type: left hand
          1,                       // hand_device_id: hand device ID
          0,                       // canfd_device_id: USB CANFD adapter device index
          0                        // canfd_channel_id: CAN channel index (0=can0, 1=can1)
      );
    }
    std::cout << "[Info]: Using device type: " << device_type << std::endl;
  }

  void TearDown() override {
    hand_.reset();
  }

  std::unique_ptr<agilink::omnihand::OmniHandDexUMI> hand_;
};

// Test factory method
TEST_F(OmniHandDexUMITest, CreateHand) {
  EXPECT_NE(hand_, nullptr);
}

// Test initialization
TEST_F(OmniHandDexUMITest, Init) {
  ASSERT_TRUE(hand_->Init()) << "Failed to initialize device. Check hardware connection.";
}

// Test vendor info
TEST_F(OmniHandDexUMITest, GetVendorInfo) {
  ASSERT_TRUE(hand_->Init()) << "Failed to initialize device";
  auto vendor_info = hand_->GetVendorInfo();
  std::cout << "[GetVendorInfo] Vendor Info:" << std::endl;
  std::cout << vendor_info.ToString() << std::endl;
  EXPECT_EQ(vendor_info.dof, 10);  // UMI has 10 DOF
}

// Test device info
TEST_F(OmniHandDexUMITest, GetDeviceInfo) {
  ASSERT_TRUE(hand_->Init()) << "Failed to initialize device";
  auto device_info = hand_->GetDeviceInfo();
  std::cout << "[GetDeviceInfo] Device Info:" << std::endl;
  std::cout << device_info.ToString() << std::endl;
  EXPECT_EQ(device_info.hand_device_id, 1);
}

// Test tactile sensor (UMI specific)
TEST_F(OmniHandDexUMITest, TactileSensor) {
  ASSERT_TRUE(hand_->Init()) << "Failed to initialize device";
  
  // Test all 6 sensors (UMI has: Thumb, Index, Middle, Ring, Little, Palm - no Dorsum)
  std::vector<std::pair<agilink::omnihand::Finger, std::string>> fingers = {
    {agilink::omnihand::Finger::THUMB, "Thumb"},
    {agilink::omnihand::Finger::INDEX, "Index"},
    {agilink::omnihand::Finger::MIDDLE, "Middle"},
    {agilink::omnihand::Finger::RING, "Ring"},
    {agilink::omnihand::Finger::LITTLE, "Little"},
    {agilink::omnihand::Finger::PALM, "Palm"}
  };
  
  std::cout << "[GetTactileSensorDataRaw] Reading individual sensors (unit: 1g, max: 255g):" << std::endl;
  for (const auto& [finger, name] : fingers) {
    auto tactile_data = hand_->GetTactileSensorDataRaw(finger);
    std::cout << "  " << name << " (" << tactile_data.data_.size() << " values): [";
    for (size_t i = 0; i < tactile_data.data_.size(); ++i) {
      std::cout << static_cast<int>(tactile_data.data_[i]);
      if (i < tactile_data.data_.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
    EXPECT_EQ(static_cast<unsigned char>(tactile_data.sensor_id_), static_cast<unsigned char>(finger));
  }
  
  // Test getting all tactile sensor data at once
  auto all_tactile_data = hand_->GetAllTactileSensorDataRaw();
  std::cout << "\n[GetAllTactileSensorDataRaw] All " << all_tactile_data.size() << " sensors:" << std::endl;
  for (const auto& sensor : all_tactile_data) {
    std::string finger_name;
    switch (sensor.sensor_id_) {
      case agilink::omnihand::Finger::THUMB: finger_name = "Thumb"; break;
      case agilink::omnihand::Finger::INDEX: finger_name = "Index"; break;
      case agilink::omnihand::Finger::MIDDLE: finger_name = "Middle"; break;
      case agilink::omnihand::Finger::RING: finger_name = "Ring"; break;
      case agilink::omnihand::Finger::LITTLE: finger_name = "Little"; break;
      case agilink::omnihand::Finger::PALM: finger_name = "Palm"; break;
      default: finger_name = "Unknown"; break;
    }
    std::cout << "  " << finger_name << " (" << sensor.data_.size() << " values): [";
    for (size_t i = 0; i < sensor.data_.size(); ++i) {
      std::cout << static_cast<int>(sensor.data_[i]);
      if (i < sensor.data_.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
  }
  EXPECT_GE(all_tactile_data.size(), 0);
}

// Test position query (UMI specific)
TEST_F(OmniHandDexUMITest, GetJointMotorPosi) {
  ASSERT_TRUE(hand_->Init()) << "Failed to initialize device";
  
  // Test single joint position query
  for (unsigned char joint_idx = 1; joint_idx <= agilink::omnihand::OmniHandDexUMI::kDegreesOfActiveFreedom; ++joint_idx) {
    int16_t pos = hand_->GetJointMotorPosi(joint_idx);
    std::cout << "[GetJointMotorPosi] Joint " << static_cast<int>(joint_idx) 
              << " position: " << pos << std::endl;
  }
  
  // Test invalid joint index
  int16_t invalid_pos = hand_->GetJointMotorPosi(0);
  EXPECT_EQ(invalid_pos, -1);
  invalid_pos = hand_->GetJointMotorPosi(11);
  EXPECT_EQ(invalid_pos, -1);
}

TEST_F(OmniHandDexUMITest, GetAllJointMotorPosi) {
  ASSERT_TRUE(hand_->Init()) << "Failed to initialize device";
  
  // Test all joint positions query
  std::vector<int16_t> positions = hand_->GetAllJointMotorPosi();
  std::cout << "[GetAllJointMotorPosi] All joint positions (" << positions.size() << " values): ";
  for (size_t i = 0; i < positions.size(); ++i) {
    std::cout << positions[i];
    if (i < positions.size() - 1) std::cout << ", ";
  }
  std::cout << std::endl;
  
  // Expect 10 joint positions
  EXPECT_EQ(positions.size(), agilink::omnihand::OmniHandDexUMI::kDegreesOfActiveFreedom);
}

// // Test position calibration (UMI specific)
// // Warning: These tests perform actual calibration - use with caution
// TEST_F(OmniHandDexUMITest, MinPositionCalibration) {
//   ASSERT_TRUE(hand_->Init()) << "Failed to initialize device";
  
//   // Test all joints calibration
//   hand_->SetMinPositionCalibration();
//   std::cout << "[SetMinPositionCalibration] Minimum position calibration set for all joints" << std::endl;
  
//   // Test single joint calibration (joint 1-10)
//   for (unsigned char joint_idx = 1; joint_idx <= 10; ++joint_idx) {
//     hand_->SetMinPositionCalibration(joint_idx);
//     std::cout << "[SetMinPositionCalibration] Minimum position calibration set for joint " 
//               << static_cast<int>(joint_idx) << std::endl;
//   }
// }

// TEST_F(OmniHandDexUMITest, MaxPositionCalibration) {
//   ASSERT_TRUE(hand_->Init()) << "Failed to initialize device";
  
//   // Test all joints calibration
//   hand_->SetMaxPositionCalibration();
//   std::cout << "[SetMaxPositionCalibration] Maximum position calibration set for all joints" << std::endl;
  
//   // Test single joint calibration (joint 1-10)
//   for (unsigned char joint_idx = 1; joint_idx <= 10; ++joint_idx) {
//     hand_->SetMaxPositionCalibration(joint_idx);
//     std::cout << "[SetMaxPositionCalibration] Maximum position calibration set for joint " 
//               << static_cast<int>(joint_idx) << std::endl;
//   }
// }

// Main function for gtest
int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  
  // Parse command line arguments
  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if ((arg == "-d" || arg == "--device") && i + 1 < argc) {
      std::string device_arg = argv[++i];
      if (device_arg == "zlgcan" || device_arg == "hcan") {
        g_device_type = device_arg;
      } else {
        std::cerr << "[Error]: -d value must be 'zlgcan' or 'hcan', got: " << device_arg << std::endl;
        return 1;
      }
    }
  }
  
  return RUN_ALL_TESTS();
}

File: linux/x64/python/test/test_omnihand_dex_umi.py
# Copyright (c) 2025, Agibot Co., Ltd.
# OmniHand 2025 SDK is licensed under Mulan PSL v2.

"""
Unit tests for OmniHand Dex UMI
"""

import sys
import os

# Ensure we use the installed omnihand package, not the source directory
# Remove parent directory from path to avoid importing from source
parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
if parent_dir in sys.path:
    sys.path.remove(parent_dir)

import pytest
from omnihand import OmniHandDexUMI, HandType, Finger

# Global variable to store device type (zlgcan or hcan)
# Can be set via environment variable OMNIHAND_DEVICE_TYPE or -d when running directly
DEVICE_TYPE = "zlgcan"  # Default: zlgcan

# Try to get device type from environment variable first (works with pytest)
import os
env_device = os.environ.get("OMNIHAND_DEVICE_TYPE")
if env_device is not None:
    if env_device in ["zlgcan", "hcan"]:
        DEVICE_TYPE = env_device
    else:
        print(f"[Error]: OMNIHAND_DEVICE_TYPE must be 'zlgcan' or 'hcan'")

if "-d" in sys.argv:
    # Parse -d argument when running directly with Python (not with pytest)
    # Note: -d doesn't work with pytest directly, use OMNIHAND_DEVICE_TYPE env var instead
    idx = sys.argv.index("-d")
    if idx + 1 < len(sys.argv):
        device_arg = sys.argv[idx + 1]
        if device_arg in ["zlgcan", "hcan"]:
            DEVICE_TYPE = device_arg
        else:
            print(f"[Error]: -d value must be 'zlgcan' or 'hcan'")
            sys.exit(1)
        # Remove -d and its value from sys.argv so pytest.main() doesn't see them
        sys.argv.pop(idx)
        sys.argv.pop(idx)

@pytest.fixture
def hand():
    """Create and initialize OmniHand Dex UMI instance for testing"""
    if DEVICE_TYPE == "hcan":
        hand = OmniHandDexUMI.create_hand_by_hcan(
            hand_type=HandType.LEFT,
            hand_device_id=1,
            canfd_device_id=0,
            canfd_channel_id=0
        )
    else:  # default: zlgcan
        hand = OmniHandDexUMI.create_hand_by_zlgcan(
            hand_type=HandType.LEFT,
            hand_device_id=1,
            canfd_device_id=0,
            canfd_channel_id=0
        )
    print(f"[Info]: Using device type: {DEVICE_TYPE}")
    yield hand


def test_create_hand(hand):
    """Test factory method"""
    assert hand is not None


def test_init(hand):
    """Test initialization"""
    assert hand.init(), "Failed to initialize device. Check hardware connection."


def test_get_vendor_info(hand):
    """Test vendor info"""
    assert hand.init(), "Failed to initialize device"
    vendor_info = hand.get_vendor_info()
    print(f"\n[get_vendor_info] Vendor Info:")
    print(str(vendor_info))
    assert vendor_info.dof == 10  # UMI has 10 DOF


def test_get_device_info(hand):
    """Test device info"""
    assert hand.init(), "Failed to initialize device"
    device_info = hand.get_device_info()
    print(f"\n[get_device_info] Device Info:")
    print(str(device_info))
    # Only check deviceId if request succeeded (non-zero indicates success)
    if device_info.hand_device_id != 0:
        assert device_info.hand_device_id == 1


def test_tactile_sensor_raw(hand):
    """Test tactile sensor raw data (UMI specific)"""
    assert hand.init(), "Failed to initialize device"
    
    # Test all 6 sensors (UMI has: Thumb, Index, Middle, Ring, Little, Palm - no Dorsum)
    fingers = [
        (Finger.THUMB, "Thumb"),
        (Finger.INDEX, "Index"),
        (Finger.MIDDLE, "Middle"),
        (Finger.RING, "Ring"),
        (Finger.LITTLE, "Little"),
        (Finger.PALM, "Palm"),
    ]
    
    print(f"\n[get_tactile_sensor_data_raw] Reading individual sensors (unit: 1g, max: 255g):")
    for finger, name in fingers:
        tactile_data = hand.get_tactile_sensor_data_raw(finger)
        print(f"  {name} ({len(tactile_data.data)} values): {[int(x) for x in tactile_data.data]}")
        assert tactile_data.sensor_id == finger
    
    # Test getting all tactile sensor data at once
    all_tactile_data = hand.get_all_tactile_sensor_data_raw()
    print(f"\n[get_all_tactile_sensor_data_raw] All {len(all_tactile_data)} sensors:")
    finger_names = {
        Finger.THUMB: "Thumb",
        Finger.INDEX: "Index",
        Finger.MIDDLE: "Middle",
        Finger.RING: "Ring",
        Finger.LITTLE: "Little",
        Finger.PALM: "Palm",
    }
    for sensor in all_tactile_data:
        name = finger_names.get(sensor.sensor_id, "Unknown")
        print(f"  {name} ({len(sensor.data)} values): {[int(x) for x in sensor.data]}")
    assert len(all_tactile_data) >= 0


def test_get_joint_position(hand):
    """Test single joint motor position query (UMI Protocol Pn3=0x13)"""
    assert hand.init(), "Failed to initialize device"
    
    # Test single joint position query for all joints
    print(f"\n[get_joint_position] Testing single joint position query:")
    for joint_idx in range(1, 11):  # joints 1-10
        pos = hand.get_joint_position(joint_idx)
        print(f"  Joint {joint_idx} position: {pos}")


def test_get_all_joint_positions(hand):
    """Test all joint motor positions query (UMI Protocol Pn3=0x13)"""
    assert hand.init(), "Failed to initialize device"
    
    # Test all joint positions query
    positions = hand.get_all_joint_positions()
    print(f"\n[get_all_joint_positions] All joint positions ({len(positions)} values): {positions}")
    
    # Expect 10 joint positions (UMI has 10 DOF)
    assert len(positions) == 10

# Position calibration tests
# Warning: These tests perform actual calibration - use with caution
# def test_min_position_calibration(hand):
#     """Test minimum position calibration (UMI Protocol Pn8=0x08)"""
#     assert hand.init(), "Failed to initialize device"
#     
#     # Test all joints calibration (sub-register 0x00)
#     hand.set_min_position_calibration()
#     print(f"\n[set_min_position_calibration] Minimum position calibration set for all joints")
#     
#     # Test single joint calibration (sub-register 0x01-0x0A)
#     for joint_idx in range(1, 11):
#         hand.set_min_position_calibration(joint_idx)
#         print(f"[set_min_position_calibration] Minimum position calibration set for joint {joint_idx}")
#
#
# def test_max_position_calibration(hand):
#     """Test maximum position calibration (UMI Protocol Pn7=0x07)"""
#     assert hand.init(), "Failed to initialize device"
#     
#     # Test all joints calibration (sub-register 0x00)
#     hand.set_max_position_calibration()
#     print(f"\n[set_max_position_calibration] Maximum position calibration set for all joints")
#     
#     # Test single joint calibration (sub-register 0x01-0x0A)
#     for joint_idx in range(1, 11):
#         hand.set_max_position_calibration(joint_idx)
#         print(f"[set_max_position_calibration] Maximum position calibration set for joint {joint_idx}")

if __name__ == "__main__":
    # Default to verbose mode if -v is not already specified
    args = sys.argv[1:]
    if "-v" not in args and "--verbose" not in args:
        args = ["-v"] + args
    pytest.main([__file__] + args)
